### PR TITLE
BLocking ESTimation-based Scheduler (BLEST)

### PR DESCRIPTION
**Pull request by dweb32**
This pull request adds the BLocking ESTimation-based Scheduler (BLEST) as a module to MPTCP v0.94. 

This scheduler works much like the default MPTCP scheduler. It always prefers the subflow with the smallest round-trip-time that is available. If the network paths are very heterogeneous, then BLEST can be more beneficial than the default scheduler because it estimates if sending on a "slower" subflow would cause head-of-line blocking and decides to wait until the "faster" subflow becomes available again.

BLEST was initially proposed and evaluated in this scientific publication:
> Ferlin, S., Alay, Ö., Mehani, O., and Boreli, R. BLEST: blocking estimation-based MPTCP scheduler for heterogeneous networks. In 2016 IFIP Networking Conference, Networking 2016 and Work- shops, Vienna, Austria, May 17-19, 2016 (2016), IEEE, pp. 431–439.

This pull request is based on the implementation that was published by the original authors for MPTCP v0.89 here: https://bitbucket.org/blest_mptcp/mptcp-paper-ifip2016

I ported the code to MPTCP v0.94 and and tried to align it as close as possible to the official MPTCP project.

Note, that BLEST in this implementation does not consider if a subflow is assigned to backup state!

### CODE REVIEW

File: include/net/mptcp.h
Comment (cpaasch) [inline include/net/mptcp.h]: This is not really needed, right?

File: net/mptcp/Kconfig
Comment (cpaasch) [inline net/mptcp/Kconfig]: Please revert these changes.

File: net/mptcp/mptcp_ctrl.c
Comment (cpaasch) [inline net/mptcp/mptcp_ctrl.c]: Why is this needed?

File: net/mptcp/mptcp_sched.c
Comment (cpaasch) [inline net/mptcp/mptcp_sched.c]: I think the alignment needs some fixing here, right?

File: net/mptcp/mptcp_sched.c
Comment (cpaasch) [inline net/mptcp/mptcp_sched.c]: Looking through the code in ```mptcp_blest.c``` it does not seem that this change is really needed, right?

**Review by cpaasch — CHANGES_REQUESTED** [review]
I did a first pass, checking only the pieces that touch regular MPTCP-code.

Please rebase your pull-request on top of mptcp_trunk as well, as that's where your changes would land.

File: include/net/mptcp.h
Comment (dweb32) [reply include/net/mptcp.h]: Yes, you are right. This value is way to big. This change was introduced for other experimental scheduler implementations. I will revert this line.

File: net/mptcp/Kconfig
Comment (dweb32) [reply net/mptcp/Kconfig]: Ok, I will revert this changes. I just tried to stick as close as possible to the coding style (section 10) mentioned here: https://www.kernel.org/doc/html/v4.19/process/coding-style.html
>  Lines under a config definition are indented with one tab, while help text is indented an additional two spaces

File: net/mptcp/mptcp_ctrl.c
Comment (dweb32) [reply net/mptcp/mptcp_ctrl.c]: When going from v0.89 to v0.90, this commit 53b2260d91a34317d650b3b300798df44fc64d98 _mptcp: Do not allocate mptcp_tcp_sock for meta_ was done to save resources, as socket-related things are not necessarily useful for the meta socket. However, BLEST uses this to store settings to apply to all subflows when scheduling. 

Therefore, this fix is needed to stop the kernel crashing with this NULL pointer exception:
```
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.191907] BUG: unable to handle
kernel NULL pointer dereference at 00000000000000a4
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.192313] IP:
blest_get_available_subflow+0x228/0x5a0 [mptcp_blest]
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.192708] PGD 0 P4D 0
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.193094] Oops: 0000 [#1] SMP PTI
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.193470] Modules linked in:
mptcp_blest mptcp_fullmesh sch_netem iptable_filter crct10dif_pclmul
crc32_pclmul ghash_clmulni_intel pcbc aesni_intel aes_x86_64 crypto_simd
glue_helper cryptd vmw_balloon intel_rapl_perf joydev evdev snd_pcm snd_timer
snd soundcore pcspkr serio_raw vmwgfx ttm drm_kms_helper sg drm shpchp ac
button vmw_vsock_vmci_transport vsock vmw_vmci ip_tables x_tables autofs4 ext4
crc32c_generic crc16 mbcache jbd2 fscrypto sr_mod cdrom ata_generic sd_mod
crc32c_intel psmouse ata_piix vmxnet3 vmw_pvscsi i2c_piix4 libata scsi_mod
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.195818] CPU: 0 PID: 5 Comm:
kworker/u2:0 Not tainted 4.14.41.mptcp #13
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.196245] Hardware name: VMware, Inc.
VMware Virtual Platform/440BX Desktop Reference Platform, BIOS 6.00 07/28/2017
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.197104] Workqueue: mptcp_wq
create_subflow_worker [mptcp_fullmesh]
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.197535] task: ffff953c7ccee140
task.stack: ffffb6e000378000
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.197969] RIP:
0010:blest_get_available_subflow+0x228/0x5a0 [mptcp_blest]
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.198399] RSP: 0018:ffffb6e00037bb98
EFLAGS: 00010287
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.198821] RAX: ffff953c7a2b2200 RBX:
0000000000000000 RCX: 0000000000000000
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.199240] RDX: 0000000000000000 RSI:
ffff953c35f21480 RDI: ffff953c7cb39180
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.199652] RBP: ffff953c7caf0140 R08:
ffff953c7a2ab140 R09: 0000000000000000
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.200071] R10: 0000000000004ea8 R11:
0000000000000001 R12: 00000000000049e4
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.200468] R13: 0000000000000000 R14:
ffff953c7cb39180 R15: ffff953c35f21480
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.200859] FS:  0000000000000000(0000)
GS:ffff953c7fc00000(0000) knlGS:0000000000000000
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.201244] CS:  0010 DS: 0000 ES: 0000
CR0: 0000000080050033
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.201621] CR2: 00000000000000a4 CR3:
000000001d80a004 CR4: 00000000007606f0
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.202021] DR0: 0000000000000000 DR1:
0000000000000000 DR2: 0000000000000000
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.202388] DR3: 0000000000000000 DR6:
00000000fffe0ff0 DR7: 0000000000000400
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.202737] PKRU: 55555554
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.203083] Call Trace:
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.203433] 
mptcp_next_segment+0x6b/0x1f0 [mptcp_blest]
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.203795]  mptcp_write_xmit+0xbf/0x410
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.204158] 
__tcp_push_pending_frames+0x38/0xd0
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.204526] 
mptcp_handle_options+0x127/0x490
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.204900] 
tcp_validate_incoming+0x3a2/0x4a0
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.205277] 
tcp_rcv_established+0x151/0x580
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.205656]  ? __schedule+0x3e1/0x870
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.206040]  tcp_v4_do_rcv+0x101/0x230
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.206416]  mptcp_backlog_rcv+0x86/0xc0
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.206784]  __release_sock+0x7c/0xd0
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.207147]  release_sock+0x2b/0x90
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.207509] 
create_subflow_worker+0x41a/0x5c0 [mptcp_fullmesh]
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.207872]  process_one_work+0x177/0x360
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.208223]  worker_thread+0x4d/0x3c0
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.208562]  kthread+0xff/0x130
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.208889]  ?
process_one_work+0x360/0x360
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.209206]  ?
kthread_create_on_node+0x70/0x70
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.209515]  ret_from_fork+0x35/0x40
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.209834] Code: 00 8b 93 c0 05 00 00
39 90 a0 00 00 00 0f 83 c5 fe ff ff 89 90 a0 00 00 00 8b 93 c0 05 00 00 e9 b4
fe ff ff 49 8b 98 e8 07 00 00 <8b> bb a4 00 00 00 85 ff 0f 84 b4 00 00 00 48 8b
85 e8 07 00 00
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.210762] RIP:
blest_get_available_subflow+0x228/0x5a0 [mptcp_blest] RSP: ffffb6e00037bb98
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.211079] CR2: 00000000000000a4
Jun 13 16:08:27 mdw-linux-03 kernel: [   15.211389] ---[ end trace
961654f78726ddff ]—
```

File: net/mptcp/mptcp_sched.c
Comment (dweb32) [reply net/mptcp/mptcp_sched.c]: I just realized that this change is no longer necessary. Looking again at the BLEST code, then it does not call `mptcp_is_temp_unavailable` directly. I forgot to remove this change when I untwisted the BLEST implementation from the DAPS, OTIAS, and LAMPS schedulers, which I also ported to v0.94.

Beyond that, you are of course right. If we remove the `static` modifier the alignment needs to be corrected. But now I think I'm just reverting that change.

File: net/mptcp/mptcp_sched.c
Comment (dweb32) [reply net/mptcp/mptcp_sched.c]: Good catch! Thank you for reviewing the code. This change is a remainder of the LAMPS scheduler (see https://github.com/lampsmptcp/lamps). In our case it is completely non functional and can be reverted.

File: include/net/mptcp.h
Comment (dweb32) [inline include/net/mptcp.h]: This line needs to be removed as well if `mptcp_is_temp_unavailable` becomes `static` again.

**Comment by dweb32** [comment]
I have applied your feedback and rebased the pull request onto `mptcp_trunk`. I needed to force push my previous commits in order to resolve some conflicts in `tcp_input.c`. The original commit introduced `mptcp_set_ca_state` in `tcp_input.c` which was not used by BLEST and removed by my later commits causing the trouble.

I don't know why GitHub is still showing merge conflicts on this page, but now it should be correctly rebased. It would be great if you could re-review the code again.

File: net/mptcp/mptcp_ctrl.c
Comment (cpaasch) [reply net/mptcp/mptcp_ctrl.c]: I see.

The thing is that for the meta-socket, we store scheduling information in the mpcb. Thus, in blestsched_modify_lambda you should rather get the priv from there instead of using mptcp_tcp_sock.

Makes sense?

**Comment by cpaasch** [comment]
Your PR doesn't compile anymore. We changed ```mptcp_for_each_sk``` to ```mptcp_for_each_sub``` with different arguments. Please test your change as well.

File: net/mptcp/mptcp_ctrl.c
Comment (dweb32) [reply net/mptcp/mptcp_ctrl.c]: I will rework the code in `mptcp_blest.c` to get the priv from the mpcb properly and remove the changes in `mptcp_ctrl.c`, because this fix should then no longer be needed.

**Comment by dweb32** [comment]
I'm sorry for overlooking the changes after rebasing the PR. I just fixed the wrong calls in `mptcp_blest.c`. But I will need some time to apply the remaining changes.

At this point I think I should also rewrite `blest_get_available_subflow` to become aware of backup-subflows as well. @sferlin, do you have any suggestions, if this would change the original algorithm design or semantics of the scheduler? Otherwise, adopting `get_subflow_from_selectors` from `mptcp_sched.c` should be fine to apply BLEST first on all active-subflows before going through the backup-subflows. In both cases, we just need to find the "global-best" and the "current-best" sks to compare their stats and decide whenever to return NULL instead of the "current-best" sk.

**Comment by dweb32** [comment]
In the meantime, I have taken a closer look at the requested changes and also tried to refactor the original BLEST implementation to align better with the default scheduler. Now I got stuck and need some advice or general design decision how to rewrite BLEST such that it fits nice into MPTCP.

The first thing to notice is that although I already removed as much redundant code as possible, there are still many code lines that are identical to `mptcp_sched.c`. This isn't very surprising, because BLEST operates exactly like minRTT in most of the cases. Thus, these functions are very similar to the default scheduler:
- `mptcp_blest_next_segment`
- `__mptcp_blest_next_segment`
- `mptcp_blest_rcv_buf_optimization`
- `blest_get_available_subflow`
- `blest_get_subflow_from_selectors`

Do you have any advice, how to get rid of this redundancy? The only true functional difference is in `blest_get_subflow_from_selectors`. One possible solution could be to add a parameter to the default scheduler to optionally enable the blocking estimation algorithm. Unfortunately, this would somehow contradict the current modular design of the schedulers.

An other problem is the size of the priv, but I don't have a solution to be more efficient yet. Maybe we split `blestsched_priv` into two structs (like it is already done in `mptcp_redundant.c`), one per subflow and one as control block for the scheduler. As of now, we have all these in one priv:
```
        u32 last_rbuf_opti;
	u32 min_srtt;
	u32 max_srtt;

	int lambda_1000;
	u32 last_lambda_update;
	u32 retrans_count;
```

I updated the PR just to discuss further changes to the code. But it won't compile at the moment with the smaller MPTCP_SCHED_SIZE. The last commit also includes changes in `blest_get_available_subflow` to be aware of backup-subflows.

File: include/net/mptcp.h
Comment (dweb32) [reply include/net/mptcp.h]: Well, 128 might be too big, but the ```blestsched_priv``` needs to be somehow optimized to become smaller.

**Comment by cpaasch** [comment]
> In the meantime, I have taken a closer look at the requested changes and also tried to refactor the original BLEST implementation to align better with the default scheduler. Now I got stuck and need some advice or general design decision how to rewrite BLEST such that it fits nice into MPTCP.
> 
> The first thing to notice is that although I already removed as much redundant code as possible, there are still many code lines that are identical to `mptcp_sched.c`. This isn't very surprising, because BLEST operates exactly like minRTT in most of the cases. Thus, these functions are very similar to the default scheduler:
> 
> * `mptcp_blest_next_segment`
> * `__mptcp_blest_next_segment`
> * `mptcp_blest_rcv_buf_optimization`
> * `blest_get_available_subflow`
> * `blest_get_subflow_from_selectors`
> 
> Do you have any advice, how to get rid of this redundancy? The only true functional difference is in `blest_get_subflow_from_selectors`. One possible solution could be to add a parameter to the default scheduler to optionally enable the blocking estimation algorithm. Unfortunately, this would somehow contradict the current modular design of the schedulers.

I am fine with having some code-duplication in blest and keep the default-scheduler untouched.

If BLEST is getting more and more popular, it can be refactored later on.

> 
> An other problem is the size of the priv, but I don't have a solution to be more efficient yet. Maybe we split `blestsched_priv` into two structs (like it is already done in `mptcp_redundant.c`), one per subflow and one as control block for the scheduler. As of now, we have all these in one priv:

Yes, splitting is probably the way to go.

> 
> ```
>         u32 last_rbuf_opti;
> 	u32 min_srtt;
> 	u32 max_srtt;
> 
> 	int lambda_1000;
> 	u32 last_lambda_update;
> 	u32 retrans_count;
> ```
> I updated the PR just to discuss further changes to the code. But it won't compile at the moment with the smaller MPTCP_SCHED_SIZE. The last commit also includes changes in `blest_get_available_subflow` to be aware of backup-subflows.

**Comment by dweb32** [comment]
Finally, I'm back in the lab and have found some time to apply the requested changes and do some testing. I'm sorry it took so long to update the PR.

I was able to refactor the original BLEST code. Now, the `MPTCP_SCHED_SIZE` no longer needs to be increased because I splitted the `blestsched_priv` into two parts and placed `bestsched_cb` into the `mpcb`. Both structs are smaller and do fit. Thus, the changes to `mptcp_ctrl.c` were removed since this fix with the extra allocation isn't necessary any more.

In addition, I also removed some code of BLEST by directly calling `get_available_subflow` of the default scheduler instead.

**Comment by matttbe** [comment]
Hi @s6dlwebe ,

That's good the changes in the other files are now minimum! It will ease the integration here!
I hope Christoph will have time to finish the review :)

If you have a bit more time, may you rebase your branch on latest `origin/mptcp_trunk` and launch `checkpatch.pl` script to uniform the code?

    ./scripts/checkpatch.pl --strict <path/to/your/patch>

(you can ignore long lines, we usually try to limit the line below ~100 chars, not the default 80)

**Comment by dweb32** [comment]
> If you have a bit more time, may you rebase your branch on latest `origin/mptcp_trunk` and launch `checkpatch.pl` script to uniform the code?

I rebased the pull request and fixed some of the styling errors. Here are the remaining problems (not including the line limits). These lines are identical with `mptcp_sched.c`, so i did not modify them.
```
./scripts/checkpatch.pl --strict -f net/mptcp/mptcp_blest.c 

CHECK: Avoid CamelCase: <TCP_CA_Open>
#299: FILE: net/mptcp/mptcp_blest.c:299:
+			if (tp->srtt_us < tp_it->srtt_us && inet_csk((struct sock *)tp_it)->icsk_ca_state == TCP_CA_Open) {

WARNING: Missing a blank line after declarations
#319: FILE: net/mptcp/mptcp_blest.c:319:
+		bool do_retrans = false;
+		mptcp_for_each_sub(tp->mpcb, mptcp) {

WARNING: else is not generally useful after a break or return
#332: FILE: net/mptcp/mptcp_blest.c:332:
+					break;
+				} else {

total: 0 errors, 32 warnings, 1 checks, 497 lines checked
```

File: net/mptcp/mptcp_blest.c
Comment (cpaasch) [inline net/mptcp/mptcp_blest.c]: Could you ```#define``` these constants 1 and 2 ?

File: net/mptcp/mptcp_blest.c
Comment (cpaasch) [inline net/mptcp/mptcp_blest.c:119]: It seems like this assumes Reno as congestion-control. Might be good to spell it out in a comment.

File: net/mptcp/mptcp_blest.c
Comment (cpaasch) [inline net/mptcp/mptcp_blest.c]: Minor: Let's rename them to ```min_srtt_us``` and ```max_srtt_us```.

File: net/mptcp/mptcp_blest.c
Comment (cpaasch) [inline net/mptcp/mptcp_blest.c]: I don't think I should be on that list ;-)

**Review by cpaasch — CHANGES_REQUESTED** [review]
Hard a few minor comments. If you can address these as well, it would be great.

Overall looks very good!

File: net/mptcp/mptcp_blest.c
Comment (dweb32) [reply net/mptcp/mptcp_blest.c]: Ok, but you were already on the list in the version I started working on 
>https://bitbucket.org/blest_mptcp/mptcp-paper-ifip2016

I kept this line untouched and think Simone put you on the list because she reused large parts from the default scheduler implementation. I just want to give proper credit to anyone who was involved. I also added a short header to the file to mention the people that initially implement and evaluated the algorithm.

For my part, I just took Simone's implementation and used it for my master's thesis. Also, I'm not sure if I belong to this list because I only refactored BLEST to match with the rest of the MPTCP project and made no functional changes to the algorithm. This is my very first "real" contribution to the open source community. So feel free to correct me if I'm giving the credits in the wrong way.

File: net/mptcp/mptcp_blest.c
Comment (dweb32) [reply net/mptcp/mptcp_blest.c]: This change makes sense.

There are some other names that I am unhappy with. But if I compare the code of the other schedulers it is hard to find a matching and consistent naming scheme.

Can you give me an advice for naming
- `bsp` matches with `dsp` and `rsp` but not the redundant scheduler that calls this `sk_data`
- `metabsp` sounds wired and could be renamed to `bsc` or `bscb` to make clear it's the control block of BLEST
- should I use a underscore for `minsk`, `bestsk`, `besttp`, etc. since I have seen both ways of writing this like `meta_sk`?

File: net/mptcp/mptcp_blest.c
Comment (dweb32) [reply net/mptcp/mptcp_blest.c]: I'm actually not sure if these constants are needed any more and could be completely removed. They are only used to disable RP. If we remove this option, this function will be identical to `mptcp_rcv_buf_optimization` from `mptcp_sched.c`.

I think that there is no general benefit in disabling RP and this piece is a remainder of the evaluation of your paper with Simone.
> Paasch, Christoph, et al. "Experimental evaluation of multipath TCP schedulers." Proceedings of the 2014 ACM SIGCOMM workshop on Capacity sharing workshop. ACM, 2014.

File: net/mptcp/mptcp_blest.c
Comment (cpaasch) [reply net/mptcp/mptcp_blest.c]: > Ok, but you were already on the list in the version I started working on
> 
> > https://bitbucket.org/blest_mptcp/mptcp-paper-ifip2016
> 
> I kept this line untouched and think Simone put you on the list because she reused large parts from the default scheduler implementation. I just want to give proper credit to anyone who was involved. I also added a short header to the file to mention the people that initially implement and evaluated the algorithm.
> 
> For my part, I just took Simone's implementation and used it for my master's thesis. Also, I'm not sure if I belong to this list because I only refactored BLEST to match with the rest of the MPTCP project and made no functional changes to the algorithm. This is my very first "real" contribution to the open source community. So feel free to correct me if I'm giving the credits in the wrong way.

I think that the work of the one who picks up the pieces and puts the code in a form that is "upstreamable", creates the pull-requests, spends time interacting with the maintainers, applying the feedback,... is as valuable as the one who did the initial prototype implementation.

Thus, you definitely deserve the credit !

File: net/mptcp/mptcp_blest.c
Comment (cpaasch) [reply net/mptcp/mptcp_blest.c]: > I'm actually not sure if these constants are needed any more and could be completely removed. They are only used to disable RP. If we remove this option, this function will be identical to `mptcp_rcv_buf_optimization` from `mptcp_sched.c`.
> 
> I think that there is no general benefit in disabling RP and this piece is a remainder of the evaluation of your paper with Simone.

Fair enough. I think it's ok to entirely remove these.

> 
> > Paasch, Christoph, et al. "Experimental evaluation of multipath TCP schedulers." Proceedings of the 2014 ACM SIGCOMM workshop on Capacity sharing workshop. ACM, 2014.

File: net/mptcp/mptcp_blest.c
Comment (cpaasch) [reply net/mptcp/mptcp_blest.c]: Yeah, the naming is not consistent. Something we should probably cleanup in a separate PR (if you want to be the one ;-))



> This change makes sense.
> 
> There are some other names that I am unhappy with. But if I compare the code of the other schedulers it is hard to find a matching and consistent naming scheme.
> 
> Can you give me an advice for naming
> 
> * `bsp` matches with `dsp` and `rsp` but not the redundant scheduler that calls this `sk_data`

I think that the redundant scheduler's naming should be fixed there.

Instead of bsp, you could also use blest_p. I actually think that dsp and rsp are not very explicit. Might be better to name them def_p, rr_p, ...

> * `metabsp` sounds wired and could be renamed to `bsc` or `bscb` to make clear it's the control block of BLEST

What about blest_cb ?

> * should I use a underscore for `minsk`, `bestsk`, `besttp`, etc. since I have seen both ways of writing this like `meta_sk`?

Not needed.


All of that being said, you don't need to spend time renaming things. The PR is good as it is. Let me know if you want to do one more iteration, or not.

Comment (dweb32) [reply net/mptcp/mptcp_blest.c]: I think I'm done with the renaming. Should I just squash all commits into one and sign it?

File: net/mptcp/mptcp_blest.c
Comment (dweb32) [reply net/mptcp/mptcp_blest.c:119]: Left a comment and hope it's somehow clearer now.

File: net/mptcp/mptcp_blest.c
Comment (dweb32) [reply net/mptcp/mptcp_blest.c]: The option to disable RP is now completely removed.

File: net/mptcp/mptcp_blest.c
Comment (matttbe) [reply net/mptcp/mptcp_blest.c]: > Should I just squash all commits into one and sign it?

Thanks for having done that! :)

**Review by cpaasch — APPROVED** [review]

**Comment by cpaasch** [comment]
Looks good to me! @matttbe, you want to give your thumbs up as well?

File: net/mptcp/mptcp_blest.c
Comment (matttbe) [inline net/mptcp/mptcp_blest.c]: detail: there are macros for that: `U32_MAX` (same below)

**Review by matttbe — COMMENTED** [review]
> Looks good to me! @matttbe, you want to give your thumbs up as well?

I didn't look in details but it seems good to me! :+1:

File: net/mptcp/mptcp_blest.c
Comment (dweb32) [reply net/mptcp/mptcp_blest.c]: > detail: there are macros for that: `U32_MAX` (same below)

Thanks for the hint @matttbe ! I gonna use the macro instead. Please tell me if I should do a seperate commit once @cpaasch has already approved the PR or perform a force-push?

Thanks for your Review :)

Comment (matttbe) [reply net/mptcp/mptcp_blest.c]: if it is not merged yet, feel free to override this commit  ;-)

Comment (dweb32) [reply net/mptcp/mptcp_blest.c]: > feel free to override this commit ;-)

@matttbe I have overridden the commit now to use the `U32_MAX`. It now also includes a fix for the compiler warning that `minsk` could be uninitialized.

Comment (matttbe) [reply net/mptcp/mptcp_blest.c]: good catch!
Thx for the update!

https://github.com/multipath-tcp/mptcp/compare/91a73ec36ea427307f5cf55d02f7474fd455fa2f..e34cb523a8ca561d1cef8d85c23f90295d08ef13

**Review by matttbe — APPROVED** [review]

**Comment by matttbe** [comment]
@cpaasch : I hope you don't mind if I merge this. You already said it was OK for you ;-)

@s6dlwebe again, thank you for your work!

**Comment by dweb32** [comment]
@matttbe I noticed that the commit signature is now broken since you pushed to `mptcp_tunk`. Is this right?

**Comment by matttbe** [comment]
@s6dlwebe yes, that's because I asked Github not to create a merge commit (we don't do that for MPTCP-related changes) but to rebase on top of `mptcp_trunk`. I thought your commit was already on top of `mptcp_trunk` or maybe it was but it seems Github modified it.

Next time you create a patch, I will apply it like I usually do when patches come from the ML and sign it.

**Pull request merged**
